doc/source/whatsnew/v1.2.0.rst

```diff
@@ -305,6 +305,7 @@ Indexing
 Missing
 ^^^^^^^
 
+- Bug in :meth:`DataFrame.mask` and :meth:`DataFrame.where` raising an ``AssertionError`` when using a nullable boolean mask (:issue:`35429`)
```

Comment:
> This is taking the approach of treating NA as False within NDFrame.mask and NDFrame.where when dealing with a nullable boolean mask.

hmm, thinking some more...

from https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.DataFrame.mask.html

> Where cond is False, keep the original value. Where True, replace with corresponding value from other.

and 

```
>>> bool(pd.NA)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas\_libs\missing.pyx", line 360, in pandas._libs.missing.NAType.__bool__
    raise TypeError("boolean value of NA is ambiguous")
TypeError: boolean value of NA is ambiguous
>>>
```

maybe this isn't a bug and this should raise, but not AssertionError, see https://github.com/pandas-dev/pandas/issues/35429#issuecomment-668087893

otherwise I think the docs should to be updated to be explicit about the behaviour.

We could be in danger here of introducing a behaviour that we would want to deprecate in the future.

Reply:
i tend to agree with this

Reply:
Because mask / where are used in other places that would mean those other cases are left broken as well, e.g., https://github.com/pandas-dev/pandas/issues/36395. It also feels natural that `s[cond] += 999` be equivalent to `s = s.mask(cond, 999)` and on master:
```python
[ins] In [16]: s
Out[16]: 
0    1
1    2
2    3
dtype: int64

[ins] In [17]: cond
Out[17]: 
0     True
1    False
2     <NA>
dtype: boolean

[ins] In [18]: s[cond] = 999

[ins] In [19]: s
Out[19]: 
0    999
1      2
2      3
dtype: int64
```

Reply:
The example in #36395 looks like its trying to use `np.array([[True, pd.NA]], dtype=object)` as a boolean mask, which i think _should_ raise (thought not AssertionError)

(with 2D EAs we can make this a little more nicely behaved and have `BooleanArray([[True, pd.NA]])`) and get a nicer exception message, but it still isn't clear that should be allowed

Reply:
I think allowing indexing with nullable booleans is more useful than raising personally. This already works for single 1D arrays (the same way it would when filtering with a nullable boolean column in SQL):
```python
[ins] In [11]: s
Out[11]: 
0    1
1    2
2    3
dtype: Int64

[ins] In [12]: mask
Out[12]: 
0     True
1    False
2     <NA>
dtype: boolean

[ins] In [13]: s[mask]
Out[13]: 
0    1
dtype: Int64
```
The argument for allowing it for mask / where as well for me seems to be that these are very much like indexing operations, for which it's useful to treat as above.